File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String>result = new ArrayList<>();
        int open=0, closed=0;
        backtrack( n,open, closed, new StringBuilder(), result);
        return result;
    }
    public void backtrack(int n,int open, int closed, StringBuilder sb, List<String>result){
        if(sb.length()==2*n){
            result.add(sb.toString());
            return ;
        }
        if(open<n){
            sb.append("(");
            backtrack( n, open+1, closed, sb, result);
            sb.deleteCharAt(sb.length()-1);
        }
         if(closed<open){
            sb.append(")");
            backtrack( n, open, closed+1, sb, result);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string>ans;
        backtrack(n,0,0,ans,"");
        return ans;
    }
    void backtrack(int n,int open, int closed, vector<string>&ans, string res){
        if(res.length()==2*n){
            ans.push_back(res);
            return;
        }
        if( open <n){
            backtrack(n, open+1, closed, ans, res+"(");
        }
         if(closed < open){
            backtrack(n, open, closed+1, ans, res+")");
        }
    }
};